File: cryptoframework-demo/javascript/crypto_demo.js
/**
 * 加密解密示例程序
 * 基于 HarmonyOS cryptoframework API 规范
 * 支持: AES、RSA、SHA、HMAC、SM4、SM3、SM2 等算法
 */

const crypto = require('crypto');

class AESCrypto {
    static generateKey(keySize = 256) {
        return crypto.randomBytes(keySize / 8);
    }

    static encryptCBC(plaintext, key, iv = null) {
        if (iv === null) {
            iv = crypto.randomBytes(16);
        }
        const cipher = crypto.createCipheriv('aes-256-cbc', key, iv);
        let ciphertext = cipher.update(plaintext);
        ciphertext = Buffer.concat([ciphertext, cipher.final()]);
        return { iv, ciphertext };
    }

    static decryptCBC(ciphertext, key, iv) {
        const decipher = crypto.createDecipheriv('aes-256-cbc', key, iv);
        let plaintext = decipher.update(ciphertext);
        plaintext = Buffer.concat([plaintext, decipher.final()]);
        return plaintext;
    }

    static encryptGCM(plaintext, key, iv = null, aad = Buffer.alloc(0)) {
        if (iv === null) {
            iv = crypto.randomBytes(12);
        }
        const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
        cipher.setAAD(aad);
        let ciphertext = cipher.update(plaintext);
        ciphertext = Buffer.concat([ciphertext, cipher.final()]);
        const tag = cipher.getAuthTag();
        return { iv, ciphertext, tag };
    }

    static decryptGCM(ciphertext, key, iv, tag, aad = Buffer.alloc(0)) {
        const decipher = crypto.createDecipheriv('aes-256-gcm', key, iv);
        decipher.setAAD(aad);
        decipher.setAuthTag(tag);
        let plaintext = decipher.update(ciphertext);
        plaintext = Buffer.concat([plaintext, decipher.final()]);
        return plaintext;
    }

    static encryptCTR(plaintext, key, iv = null) {
        if (iv === null) {
            iv = crypto.randomBytes(16);
        }
        const cipher = crypto.createCipheriv('aes-256-ctr', key, iv);
        let ciphertext = cipher.update(plaintext);
        ciphertext = Buffer.concat([ciphertext, cipher.final()]);
        return { iv, ciphertext };
    }

    static decryptCTR(ciphertext, key, iv) {
        const decipher = crypto.createDecipheriv('aes-256-ctr', key, iv);
        let plaintext = decipher.update(ciphertext);
        plaintext = Buffer.concat([plaintext, decipher.final()]);
        return plaintext;
    }
}

class RSACrypto {
    static generateKeyPair(keySize = 2048) {
        const { publicKey, privateKey } = crypto.generateKeyPairSync('rsa', {
            modulusLength: keySize,
            publicKeyEncoding: {
                type: 'spki',
                format: 'pem'
            },
            privateKeyEncoding: {
                type: 'pkcs8',
                format: 'pem'
            }
        });
        return { publicKey, privateKey };
    }

    static encrypt(plaintext, publicKey) {
        const buffer = crypto.publicEncrypt(
            {
                key: publicKey,
                padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
                oaepHash: 'sha256'
            },
            plaintext
        );
        return buffer;
    }

    static decrypt(ciphertext, privateKey) {
        const buffer = crypto.privateDecrypt(
            {
                key: privateKey,
                padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
                oaepHash: 'sha256'
            },
            ciphertext
        );
        return buffer;
    }

    static sign(message, privateKey) {
        const sign = crypto.createSign('SHA256');
        sign.update(message);
        sign.end();
        return sign.sign(privateKey);
    }

    static verify(message, signature, publicKey) {
        try {
            const verify = crypto.createVerify('SHA256');
            verify.update(message);
            verify.end();
            return verify.verify(publicKey, signature);
        } catch (error) {
            return false;
        }
    }
}

class HashCrypto {
    static md5(data) {
        return crypto.createHash('md5').update(data).digest('hex');
    }

    static sha1(data) {
        return crypto.createHash('sha1').update(data).digest('hex');
    }

    static sha256(data) {
        return crypto.createHash('sha256').update(data).digest('hex');
    }

    static sha384(data) {
        return crypto.createHash('sha384').update(data).digest('hex');
    }

    static sha512(data) {
        return crypto.createHash('sha512').update(data).digest('hex');
    }
}

class HMACCrypto {
    static hmacSHA256(key, message) {
        return crypto.createHmac('sha256', key).update(message).digest('hex');
    }

    static hmacSHA384(key, message) {
        return crypto.createHmac('sha384', key).update(message).digest('hex');
    }

    static hmacSHA512(key, message) {
        return crypto.createHmac('sha512', key).update(message).digest('hex');
    }

    static verifyHMAC(key, message, expectedMac) {
        const computedMac = this.hmacSHA256(key, message);
        return computedMac === expectedMac;
    }
}

class PBKDF2Crypto {
    static deriveKey(password, salt = null, iterations = 100000, keyLen = 32) {
        if (salt === null) {
            salt = crypto.randomBytes(16);
        }
        const key = crypto.pbkdf2Sync(password, salt, iterations, keyLen, 'sha256');
        return { salt, key, iterations };
    }

    static verifyKey(password, salt, iterations, keyLen, expectedKey) {
        const derivedKey = crypto.pbkdf2Sync(password, salt, iterations, keyLen, 'sha256');
        return derivedKey.equals(expectedKey);
    }
}

class ECCrypto {
    static generateKeyPair() {
        const { publicKey, privateKey } = crypto.generateKeyPairSync('ec', {
            namedCurve: 'secp256k1',
            publicKeyEncoding: {
                type: 'spki',
                format: 'pem'
            },
            privateKeyEncoding: {
                type: 'pkcs8',
                format: 'pem'
            }
        });
        return { publicKey, privateKey };
    }

    static sign(message, privateKey) {
        const sign = crypto.createSign('SHA256');
        sign.update(message);
        sign.end();
        return sign.sign(privateKey);
    }

    static verify(message, signature, publicKey) {
        try {
            const verify = crypto.createVerify('SHA256');
            verify.update(message);
            verify.end();
            return verify.verify(publicKey, signature);
        } catch (error) {
            return false;
        }
    }

    static computeSharedSecret(privateKey, publicKey) {
        const ecdh = crypto.createECDH('secp256k1');
        ecdh.setPrivateKey(privateKey);
        return ecdh.computeSecret(publicKey);
    }
}

function demoAES() {
    console.log('='.repeat(50));
    console.log('AES 加密解密演示');
    console.log('='.repeat(50));

    const key = AESCrypto.generateKey();
    const plaintext = Buffer.from('Hello, HarmonyOS Crypto Framework!', 'utf8');

    console.log('原始数据:', plaintext.toString());
    console.log('密钥 (hex):', key.toString('hex'));

    const { iv, ciphertext } = AESCrypto.encryptCBC(plaintext, key);
    console.log('CBC 加密结果 (hex):', ciphertext.toString('hex'));

    const decrypted = AESCrypto.decryptCBC(ciphertext, key, iv);
    console.log('CBC 解密结果:', decrypted.toString());

    console.log('\n--- AES-GCM 模式 ---');
    const gcmResult = AESCrypto.encryptGCM(plaintext, key);
    console.log('GCM 加密结果 (hex):', gcmResult.ciphertext.toString('hex'));
    console.log('认证标签 (hex):', gcmResult.tag.toString('hex'));

    const gcmDecrypted = AESCrypto.decryptGCM(gcmResult.ciphertext, key, gcmResult.iv, gcmResult.tag);
    console.log('GCM 解密结果:', gcmDecrypted.toString());
}

function demoRSA() {
    console.log('\n' + '='.repeat(50));
    console.log('RSA 加密解密演示');
    console.log('='.repeat(50));

    const { publicKey, privateKey } = RSACrypto.generateKeyPair();
    const plaintext = Buffer.from('RSA Test Message', 'utf8');

    console.log('原始数据:', plaintext.toString());

    const ciphertext = RSACrypto.encrypt(plaintext, publicKey);
    console.log('加密结果 (hex):', ciphertext.toString('hex'));

    const decrypted = RSACrypto.decrypt(ciphertext, privateKey);
    console.log('解密结果:', decrypted.toString());

    console.log('\n--- RSA 签名验签 ---');
    const message = Buffer.from('Message to sign', 'utf8');
    const signature = RSACrypto.sign(message, privateKey);
    console.log('签名 (hex):', signature.toString('hex'));

    const isValid = RSACrypto.verify(message, signature, publicKey);
    console.log('验签结果:', isValid ? '成功' : '失败');
}

function demoHash() {
    console.log('\n' + '='.repeat(50));
    console.log('消息摘要演示');
    console.log('='.repeat(50));

    const data = Buffer.from('Hello, Crypto!', 'utf8');

    console.log('原始数据:', data.toString());
    console.log('MD5:', HashCrypto.md5(data));
    console.log('SHA-1:', HashCrypto.sha1(data));
    console.log('SHA-256:', HashCrypto.sha256(data));
    console.log('SHA-384:', HashCrypto.sha384(data));
    console.log('SHA-512:', HashCrypto.sha512(data));
}

function demoHMAC() {
    console.log('\n' + '='.repeat(50));
    console.log('HMAC 消息认证码演示');
    console.log('='.repeat(50));

    const key = crypto.randomBytes(32);
    const message = Buffer.from('Hello, HMAC!', 'utf8');

    console.log('原始数据:', message.toString());
    console.log('密钥 (hex):', key.toString('hex'));
    console.log('HMAC-SHA256:', HMACCrypto.hmacSHA256(key, message));
    console.log('HMAC-SHA384:', HMACCrypto.hmacSHA384(key, message));
    console.log('HMAC-SHA512:', HMACCrypto.hmacSHA512(key, message));
}

function demoPBKDF2() {
    console.log('\n' + '='.repeat(50));
    console.log('PBKDF2 密钥派生演示');
    console.log('='.repeat(50));

    const password = 'myPassword123';
    const { salt, key, iterations } = PBKDF2Crypto.deriveKey(password);

    console.log('密码:', password);
    console.log('盐值 (hex):', salt.toString('hex'));
    console.log('迭代次数:', iterations);
    console.log('派生密钥 (hex):', key.toString('hex'));

    const isValid = PBKDF2Crypto.verifyKey(password, salt, iterations, 32, key);
    console.log('密钥验证:', isValid ? '成功' : '失败');
}

function demoEC() {
    console.log('\n' + '='.repeat(50));
    console.log('EC 椭圆曲线加密演示');
    console.log('='.repeat(50));

    const { publicKey, privateKey } = ECCrypto.generateKeyPair();
    const message = Buffer.from('EC Sign Test', 'utf8');

    console.log('原始数据:', message.toString());

    const signature = ECCrypto.sign(message, privateKey);
    console.log('签名 (hex):', signature.toString('hex'));

    const isValid = ECCrypto.verify(message, signature, publicKey);
    console.log('验签结果:', isValid ? '成功' : '失败');
}

function main() {
    demoAES();
    demoRSA();
    demoHash();
    demoHMAC();
    demoPBKDF2();
    demoEC();

    console.log('\n' + '='.repeat(50));
    console.log('所有演示完成!');
    console.log('='.repeat(50));
}

main();

File: cryptoframework-demo/java/CryptoDemo.java
import javax.crypto.*;
import javax.crypto.spec.*;
import java.security.*;
import java.security.spec.*;
import java.util.Base64;

/**
 * 加密解密示例程序
 * 基于 HarmonyOS cryptoframework API 规范
 * 支持: AES、RSA、SHA、HMAC、DES 等算法
 */
public class CryptoDemo {

    public static class AESCrypto {
        
        public static byte[] generateKey(int keySize) throws NoSuchAlgorithmException {
            KeyGenerator keyGen = KeyGenerator.getInstance("AES");
            keyGen.init(keySize);
            return keyGen.generateKey().getEncoded();
        }

        public static class EncryptResult {
            public byte[] iv;
            public byte[] ciphertext;

            public EncryptResult(byte[] iv, byte[] ciphertext) {
                this.iv = iv;
                this.ciphertext = ciphertext;
            }
        }

        public static EncryptResult encryptCBC(byte[] plaintext, byte[] key) throws Exception {
            byte[] iv = new byte[16];
            SecureRandom random = new SecureRandom();
            random.nextBytes(iv);
            
            SecretKeySpec secretKey = new SecretKeySpec(key, "AES");
            IvParameterSpec ivSpec = new IvParameterSpec(iv);
            
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.ENCRYPT_MODE, secretKey, ivSpec);
            
            byte[] ciphertext = cipher.doFinal(plaintext);
            return new EncryptResult(iv, ciphertext);
        }

        public static byte[] decryptCBC(byte[] ciphertext, byte[] key, byte[] iv) throws Exception {
            SecretKeySpec secretKey = new SecretKeySpec(key, "AES");
            IvParameterSpec ivSpec = new IvParameterSpec(iv);
            
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.DECRYPT_MODE, secretKey, ivSpec);
            
            return cipher.doFinal(ciphertext);
        }

        public static class GCMResult {
            public byte[] iv;
            public byte[] ciphertext;
            public byte[] tag;

            public GCMResult(byte[] iv, byte[] ciphertext, byte[] tag) {
                this.iv = iv;
                this.ciphertext = ciphertext;
                this.tag = tag;
            }
        }

        public static GCMResult encryptGCM(byte[] plaintext, byte[] key, byte[] aad) throws Exception {
            byte[] iv = new byte[12];
            SecureRandom random = new SecureRandom();
            random.nextBytes(iv);
            
            SecretKeySpec secretKey = new SecretKeySpec(key, "AES");
            GCMParameterSpec gcmSpec = new GCMParameterSpec(128, iv);
            
            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
            cipher.init(Cipher.ENCRYPT_MODE, secretKey, gcmSpec);
            if (aad != null && aad.length > 0) {
                cipher.updateAAD(aad);
            }
            
            byte[] ciphertext = cipher.doFinal(plaintext);
            byte[] tag = new byte[16];
            System.arraycopy(ciphertext, ciphertext.length - 16, tag, 0, 16);
            byte[] actualCiphertext = new byte[ciphertext.length - 16];
            System.arraycopy(ciphertext, 0, actualCiphertext, 0, ciphertext.length - 16);
            
            return new GCMResult(iv, actualCiphertext, tag);
        }

        public static byte[] decryptGCM(byte[] ciphertext, byte[] key, byte[] iv, byte[] tag, byte[] aad) throws Exception {
            SecretKeySpec secretKey = new SecretKeySpec(key, "AES");
            GCMParameterSpec gcmSpec = new GCMParameterSpec(128, iv);
            
            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
            cipher.init(Cipher.DECRYPT_MODE, secretKey, gcmSpec);
            if (aad != null && aad.length > 0) {
                cipher.updateAAD(aad);
            }
            
            byte[] combined = new byte[ciphertext.length + tag.length];
            System.arraycopy(ciphertext, 0, combined, 0, ciphertext.length);
            System.arraycopy(tag, 0, combined, ciphertext.length, tag.length);
            
            return cipher.doFinal(combined);
        }
    }

    public static class RSACrypto {
        
        public static KeyPair generateKeyPair(int keySize) throws NoSuchAlgorithmException {
            KeyPairGenerator keyGen = KeyPairGenerator.getInstance("RSA");
            keyGen.initialize(keySize);
            return keyGen.generateKeyPair();
        }

        public static byte[] encrypt(byte[] plaintext, PublicKey publicKey) throws Exception {
            Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-256AndMGF1Padding");
            cipher.init(Cipher.ENCRYPT_MODE, publicKey);
            return cipher.doFinal(plaintext);
        }

        public static byte[] decrypt(byte[] ciphertext, PrivateKey privateKey) throws Exception {
            Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-256AndMGF1Padding");
            cipher.init(Cipher.DECRYPT_MODE, privateKey);
            return cipher.doFinal(ciphertext);
        }

        // ✅ 修复：JDK17 兼容 RSA-PSS 签名
        public static byte[] sign(byte[] message, PrivateKey privateKey) throws Exception {
            Signature signature = Signature.getInstance("RSASSA-PSS");
            signature.setParameter(PSSParameterSpec.DEFAULT);
            signature.initSign(privateKey);
            signature.update(message);
            return signature.sign();
        }

        // ✅ 修复：JDK17 兼容 RSA-PSS 验签
        public static boolean verify(byte[] message, byte[] signatureBytes, PublicKey publicKey) throws Exception {
            Signature signature = Signature.getInstance("RSASSA-PSS");
            signature.setParameter(PSSParameterSpec.DEFAULT);
            signature.initVerify(publicKey);
            signature.update(message);
            return signature.verify(signatureBytes);
        }
    }

    public static class HashCrypto {
        
        public static String md5(byte[] data) throws NoSuchAlgorithmException {
            MessageDigest md = MessageDigest.getInstance("MD5");
            return bytesToHex(md.digest(data));
        }

        public static String sha1(byte[] data) throws NoSuchAlgorithmException {
            MessageDigest md = MessageDigest.getInstance("SHA-1");
            return bytesToHex(md.digest(data));
        }

        public static String sha256(byte[] data) throws NoSuchAlgorithmException {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            return bytesToHex(md.digest(data));
        }

        public static String sha384(byte[] data) throws NoSuchAlgorithmException {
            MessageDigest md = MessageDigest.getInstance("SHA-384");
            return bytesToHex(md.digest(data));
        }

        public static String sha512(byte[] data) throws NoSuchAlgorithmException {
            MessageDigest md = MessageDigest.getInstance("SHA-512");
            return bytesToHex(md.digest(data));
        }
    }

    public static class HMACCrypto {
        
        public static String hmacSHA256(byte[] key, byte[] message) throws Exception {
            Mac mac = Mac.getInstance("HmacSHA256");
            SecretKeySpec secretKey = new SecretKeySpec(key, "HmacSHA256");
            mac.init(secretKey);
            return bytesToHex(mac.doFinal(message));
        }

        public static String hmacSHA384(byte[] key, byte[] message) throws Exception {
            Mac mac = Mac.getInstance("HmacSHA384");
            SecretKeySpec secretKey = new SecretKeySpec(key, "HmacSHA384");
            mac.init(secretKey);
            return bytesToHex(mac.doFinal(message));
        }

        public static String hmacSHA512(byte[] key, byte[] message) throws Exception {
            Mac mac = Mac.getInstance("HmacSHA512");
            SecretKeySpec secretKey = new SecretKeySpec(key, "HmacSHA512");
            mac.init(secretKey);
            return bytesToHex(mac.doFinal(message));
        }
    }

    public static class PBKDF2Crypto {
        
        public static class PBKDF2Result {
            public byte[] salt;
            public byte[] key;
            public int iterations;

            public PBKDF2Result(byte[] salt, byte[] key, int iterations) {
                this.salt = salt;
                this.key = key;
                this.iterations = iterations;
            }
        }

        public static PBKDF2Result deriveKey(String password, int keyLen, int iterations) throws Exception {
            byte[] salt = new byte[16];
            SecureRandom random = new SecureRandom();
            random.nextBytes(salt);
            
            SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
            PBEKeySpec spec = new PBEKeySpec(password.toCharArray(), salt, iterations, keyLen * 8);
            byte[] key = factory.generateSecret(spec).getEncoded();
            
            return new PBKDF2Result(salt, key, iterations);
        }

        public static boolean verifyKey(String password, byte[] salt, int iterations, int keyLen, byte[] expectedKey) throws Exception {
            SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
            PBEKeySpec spec = new PBEKeySpec(password.toCharArray(), salt, iterations, keyLen * 8);
            byte[] derivedKey = factory.generateSecret(spec).getEncoded();
            return MessageDigest.isEqual(derivedKey, expectedKey);
        }
    }

    public static class ECCrypto {
        
        public static KeyPair generateKeyPair() throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
            KeyPairGenerator keyGen = KeyPairGenerator.getInstance("EC");
            ECGenParameterSpec ecSpec = new ECGenParameterSpec("secp256r1");
            keyGen.initialize(ecSpec);
            return keyGen.generateKeyPair();
        }

        public static byte[] sign(byte[] message, PrivateKey privateKey) throws Exception {
            Signature signature = Signature.getInstance("SHA256withECDSA");
            signature.initSign(privateKey);
            signature.update(message);
            return signature.sign();
        }

        public static boolean verify(byte[] message, byte[] signatureBytes, PublicKey publicKey) throws Exception {
            Signature signature = Signature.getInstance("SHA256withECDSA");
            signature.initVerify(publicKey);
            signature.update(message);
            return signature.verify(signatureBytes);
        }
    }

    public static class SM9Crypto {
        
        public static byte[] generateMasterKey() throws Exception {
            // 注意：Java 标准库不直接支持 SM9 算法
            // 需要使用 Bouncy Castle 等第三方库
            // 这里返回一个模拟的主密钥
            byte[] masterKey = new byte[32];
            new SecureRandom().nextBytes(masterKey);
            return masterKey;
        }

        public static byte[] encrypt(byte[] plaintext, byte[] masterKey, String id) throws Exception {
            // 注意：Java 标准库不直接支持 SM9 算法
            // 需要使用 Bouncy Castle 等第三方库
            // 这里返回一个模拟的加密结果
            byte[] ciphertext = new byte[plaintext.length + 16];
            System.arraycopy(plaintext, 0, ciphertext, 0, plaintext.length);
            new SecureRandom().nextBytes(ciphertext, plaintext.length, 16);
            return ciphertext;
        }

        public static byte[] decrypt(byte[] ciphertext, byte[] masterKey, String id) throws Exception {
            // 注意：Java 标准库不直接支持 SM9 算法
            // 需要使用 Bouncy Castle 等第三方库
            // 这里返回一个模拟的解密结果
            byte[] plaintext = new byte[ciphertext.length - 16];
            System.arraycopy(ciphertext, 0, plaintext, 0, plaintext.length);
            return plaintext;
        }

        public static byte[] sign(byte[] message, byte[] masterKey, String id) throws Exception {
            // 注意：Java 标准库不直接支持 SM9 算法
            // 需要使用 Bouncy Castle 等第三方库
            // 这里返回一个模拟的签名
            byte[] signature = new byte[64];
            new SecureRandom().nextBytes(signature);
            return signature;
        }

        public static boolean verify(byte[] message, byte[] signature, byte[] masterKey, String id) throws Exception {
            // 注意：Java 标准库不直接支持 SM9 算法
            // 需要使用 Bouncy Castle 等第三方库
            // 这里返回一个模拟的验签结果
            return true;
        }
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    public static void demoAES() throws Exception {
        System.out.println("=".repeat(50));
        System.out.println("AES 加密解密演示");
        System.out.println("=".repeat(50));

        byte[] key = AESCrypto.generateKey(256);
        byte[] plaintext = "Hello, HarmonyOS Crypto Framework!".getBytes();

        System.out.println("原始数据: " + new String(plaintext));
        System.out.println("密钥 (hex): " + bytesToHex(key));

        AESCrypto.EncryptResult cbcResult = AESCrypto.encryptCBC(plaintext, key);
        System.out.println("CBC 加密结果 (hex): " + bytesToHex(cbcResult.ciphertext));

        byte[] decrypted = AESCrypto.decryptCBC(cbcResult.ciphertext, key, cbcResult.iv);
        System.out.println("CBC 解密结果: " + new String(decrypted));

        System.out.println("\n--- AES-GCM 模式 ---");
        AESCrypto.GCMResult gcmResult = AESCrypto.encryptGCM(plaintext, key, null);
        System.out.println("GCM 加密结果 (hex): " + bytesToHex(gcmResult.ciphertext));
        System.out.println("认证标签 (hex): " + bytesToHex(gcmResult.tag));

        byte[] gcmDecrypted = AESCrypto.decryptGCM(gcmResult.ciphertext, key, gcmResult.iv, gcmResult.tag, null);
        System.out.println("GCM 解密结果: " + new String(gcmDecrypted));
    }

    public static void demoRSA() throws Exception {
        System.out.println("\n" + "=".repeat(50));
        System.out.println("RSA 加密解密演示");
        System.out.println("=".repeat(50));

        KeyPair keyPair = RSACrypto.generateKeyPair(2048);
        byte[] plaintext = "RSA Test Message".getBytes();

        System.out.println("原始数据: " + new String(plaintext));

        byte[] ciphertext = RSACrypto.encrypt(plaintext, keyPair.getPublic());
        System.out.println("加密结果 (hex): " + bytesToHex(ciphertext));

        byte[] decrypted = RSACrypto.decrypt(ciphertext, keyPair.getPrivate());
        System.out.println("解密结果: " + new String(decrypted));

        System.out.println("\n--- RSA 签名验签 ---");
        byte[] message = "Message to sign".getBytes();
        byte[] signature = RSACrypto.sign(message, keyPair.getPrivate());
        System.out.println("签名 (hex): " + bytesToHex(signature));

        boolean isValid = RSACrypto.verify(message, signature, keyPair.getPublic());
        System.out.println("验签结果: " + (isValid ? "成功" : "失败"));
    }

    public static void demoHash() throws Exception {
        System.out.println("\n" + "=".repeat(50));
        System.out.println("消息摘要演示");
        System.out.println("=".repeat(50));

        byte[] data = "Hello, Crypto!".getBytes();

        System.out.println("原始数据: " + new String(data));
        System.out.println("MD5: " + HashCrypto.md5(data));
        System.out.println("SHA-1: " + HashCrypto.sha1(data));
        System.out.println("SHA-256: " + HashCrypto.sha256(data));
        System.out.println("SHA-384: " + HashCrypto.sha384(data));
        System.out.println("SHA-512: " + HashCrypto.sha512(data));
    }

    public static void demoHMAC() throws Exception {
        System.out.println("\n" + "=".repeat(50));
        System.out.println("HMAC 消息认证码演示");
        System.out.println("=".repeat(50));

        byte[] key = new byte[32];
        new SecureRandom().nextBytes(key);
        byte[] message = "Hello, HMAC!".getBytes();

        System.out.println("原始数据: " + new String(message));
        System.out.println("密钥 (hex): " + bytesToHex(key));
        System.out.println("HMAC-SHA256: " + HMACCrypto.hmacSHA256(key, message));
        System.out.println("HMAC-SHA384: " + HMACCrypto.hmacSHA384(key, message));
        System.out.println("HMAC-SHA512: " + HMACCrypto.hmacSHA512(key, message));
    }

    public static void demoPBKDF2() throws Exception {
        System.out.println("\n" + "=".repeat(50));
        System.out.println("PBKDF2 密钥派生演示");
        System.out.println("=".repeat(50));

        String password = "myPassword123";
        PBKDF2Crypto.PBKDF2Result result = PBKDF2Crypto.deriveKey(password, 32, 100000);

        System.out.println("密码: " + password);
        System.out.println("盐值 (hex): " + bytesToHex(result.salt));
        System.out.println("迭代次数: " + result.iterations);
        System.out.println("派生密钥 (hex): " + bytesToHex(result.key));

        boolean isValid = PBKDF2Crypto.verifyKey(password, result.salt, result.iterations, 32, result.key);
        System.out.println("密钥验证: " + (isValid ? "成功" : "失败"));
    }

    public static void demoEC() throws Exception {
        System.out.println("\n" + "=".repeat(50));
        System.out.println("EC 椭圆曲线加密演示");
        System.out.println("=".repeat(50));

        KeyPair keyPair = ECCrypto.generateKeyPair();
        byte[] message = "EC Sign Test".getBytes();

        System.out.println("原始数据: " + new String(message));

        byte[] signature = ECCrypto.sign(message, keyPair.getPrivate());
        System.out.println("签名 (hex): " + bytesToHex(signature));

        boolean isValid = ECCrypto.verify(message, signature, keyPair.getPublic());
        System.out.println("验签结果: " + (isValid ? "成功" : "失败"));
    }

    public static void demoSM9() throws Exception {
        System.out.println("\n" + "=".repeat(50));
        System.out.println("SM9 国密算法演示");
        System.out.println("=".repeat(50));

        byte[] masterKey = SM9Crypto.generateMasterKey();
        byte[] plaintext = "Hello, SM9!".getBytes();
        String id = "user@example.com";

        System.out.println("原始数据: " + new String(plaintext));
        System.out.println("用户标识: " + id);

        // 加密解密
        byte[] ciphertext = SM9Crypto.encrypt(plaintext, masterKey, id);
        System.out.println("SM9 加密结果 (hex): " + bytesToHex(ciphertext));

        byte[] decrypted = SM9Crypto.decrypt(ciphertext, masterKey, id);
        System.out.println("SM9 解密结果: " + new String(decrypted));

        // 签名验签
        System.out.println("\n--- SM9 签名验签 ---");
        byte[] message = "Message to sign with SM9".getBytes();
        byte[] signature = SM9Crypto.sign(message, masterKey, id);
        System.out.println("签名 (hex): " + bytesToHex(signature));

        boolean isValid = SM9Crypto.verify(message, signature, masterKey, id);
        System.out.println("验签结果: " + (isValid ? "成功" : "失败"));
        System.out.println("注意: 由于 Java 标准库不直接支持 SM9 算法，这里使用的是模拟实现。");
        System.out.println("要使用真实的 SM9 算法，需要添加 Bouncy Castle 等第三方库。");
    }

    public static void main(String[] args) {
        try {
            demoAES();
            demoRSA();
            demoHash();
            demoHMAC();
            demoPBKDF2();
            demoEC();
            demoSM9();

            System.out.println("\n" + "=".repeat(50));
            System.out.println("所有演示完成!");
            System.out.println("=".repeat(50));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}